### ISSUE TITLE
`Tensor.is_contiguous` is false for expanded tensors

### ISSUE THREAD

**Issue by Flamefire** [pytorch/pytorch#190038]
### 📚 The doc issue

I'm looking at https://docs.pytorch.org/docs/2.13/generated/torch.Tensor.is_contiguous.html which simply states
> Returns True if self tensor is contiguous in memory in the order specified by memory format.

However this "contiguous" is possibly ambiguous.
Background: I ran into an accuracy issue where one codepath was using an expanded tensor while the other left it unexpanded leading to different results in a matmul.
That boiled down to how "contiguous" is defined:

```
import torch
x=torch.tensor(1)
assert x.is_contiguous()
x2=x.expand((2,2))
assert not x2.is_contiguous()
assert x2.stride() == (0,0)
```

I.e. a tensor with 1 element in memory is not considered "contiguous". This matters in AoT-Autograd, e.g. in [`coerce_to_expected_memory_format`](https://github.com/pytorch/pytorch/blob/854657f330c46fb698615946874714cfca0ea5dc/torch/_functorch/_aot_autograd/runtime_wrappers.py#L2501)

Because an expanded tensor is not considered contiguous it will be materialized which leads to a difference compared to the eager path.
Example failure caused by this in `TestEagerFusionOpInfoCPU.test_aot_autograd_symbolic_exhaustive_linalg_householder_product_cpu_float32`:

> Mismatched elements: 2 / 30 (6.7%)
> Greatest absolute difference: 5.340576171875e-05 at index (5, 3) (up to 1e-05 allowed)
> Greatest relative difference: 5.707081527361879e-06 at index (5, 2) (up to 1.3e-06 allowed)

But before addressing this I'd like to properly document `is_contiguous` in a way that covers expanded tensors with all or some strides being zero. One could say that for the case of all-zero strides the tensor is still contiguous: There is no gap between elements.
Contrary a row-expanded tensor is not contiguous as after iterating over a row you'd need to jump back.

cc @svekars @sekyondaMeta @AlannaBurke @chauhang @penguinwu

**Comment by ngimel** [comment]
Expanded tensor with 0 strides is clearlly not contiguous, what exactly is the issue here?

**Comment by Flamefire** [comment]
The memory of the tensor is a contiguous chunk. So in that definition it would be. What I'd like is a short note in the docs on what contiguous means exactly here to avoid relying on conventions. E.g. a 2x3 tensor with values 1-6 would be contiguous for those 2 layouts:
`1|2|3|4|5|6` & `1|4|2|5|3|6` (row- and column major with strides (3,1) and (1,2))

~A 1x3 tensor expanded to 2x3 would be stored as `1|2|3` with strides (0,1), so the row is contiguous but iterating in C-order has a "jump" between the 3rd and 4th element. For an expanded scalar (stored as a single value with strides (0,0)) there is no movement of the pointer at all so either no "jumps" or "all jumps".~

I've just seen that https://docs.pytorch.org/docs/2.13/tensor_attributes.html#torch.memory_format clarifies this with:
> torch.contiguous_format: Tensor is or will be allocated in dense non-overlapping memory. Strides represented by values in decreasing order.

I.e. `strides[i] < strides[i+1] with strides[-1]==1` 
Interestingly no layout represents Fortran-order-contiguous.

Also the "dense non-overlapping memory." part clarifies that any stride[i]==0 makes it non-contiguous due to overlap


So if any enhancement I'd change:
```diff
- torch.contiguous_format: Tensor is or will be allocated in dense non-overlapping memory. Strides represented by values in decreasing order.
+ torch.contiguous_format: Tensor is or will be allocated in dense non-overlapping memory. Strides represented by values in strictly decreasing order, i.e. row-major-/C-layout.
```

**Comment by ngimel** [comment]
No, memory being a contiguous chunk has nothing to do with tensor being contiguous, and 2x3 tensor with values 1-6 is contiguous for exactly one layout.

**Comment by Flamefire** [comment]
> memory being a contiguous chunk has nothing to do with tensor being contiguous

That was based on definitions of contiguous used elsewhere, e.g.
> CONTIGUOUS simply means that there is no spacing between array elements. 

> and 2x3 tensor with values 1-6 is contiguous for exactly one layout.

Elsewhere there is the difference in C-contiguous and F-contiguous layouts and given the different BLAS backends that might (still) matter, although they may now be able to handle both layouts with full efficiency.

But my initial question was for the extreme case of a single value in a tensor with multiple elements stored in a single memory location. That is excluded from being contiguous by "non-overlapping memory" and "strides [...] in decreasing/increasing order", although that is weaker without requiring `strides[-1]==1` as a 1D tensor of a single value in N elements, with stride 0 would technically meet the "decreasing order". But "non-overlapping" does exclude that.

**Comment by ngimel** [comment]
"contiguous" means logical offset in a row major order matches physical offset. Expanded tensors don't satisfy this criterion
